base-deriving-via/src/Data/DerivingVia/GHC/Generics/Semigroup.hs

```diff
@@ -0,0 +1,42 @@
+{-# LANGUAGE DataKinds            #-}
+{-# LANGUAGE FlexibleContexts     #-}
+{-# LANGUAGE FlexibleInstances    #-}
+{-# LANGUAGE TypeOperators        #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+{-# OPTIONS -Wno-unticked-promoted-constructors #-}
+
+-- | "GHC.Generics" definition of '<>'
+module Data.DerivingVia.GHC.Generics.Semigroup
+  ( GSemigroup (..)
+  )
+where
+
+import GHC.Generics
+import GHC.TypeLits
+
+class GSemigroup rep where
+  gsappend :: rep x -> rep x -> rep x
+
+instance Monoid c => GSemigroup (K1 i c) where
+  gsappend (K1 l) (K1 r) = K1 (l <> r)
+
+instance GSemigroup f => GSemigroup (M1 i c f) where
+  gsappend (M1 l) (M1 r) = M1 (gsappend l r)
+
+instance GSemigroup V1 where
+  gsappend = error "GSemigroup V1"
```

Comment:
Should we use a `TypeError` here as well?

Reply:
Actually it shouldn't be an error, `Void` has a valid `Semigroup` instance.

Reply:
Thanks @coot; that was a copy-paste error. Nice catch.